core/src/main/java/io/grpc/internal/SerializingExecutor.java

```diff
@@ -82,90 +69,48 @@ public SerializingExecutor(Executor executor) {
     this.executor = executor;
   }
 
-  private final Object internalLock = new Object() {
-    @Override public String toString() {
-      return "SerializingExecutor lock: " + super.toString();
-    }
-  };
-
   /**
    * Runs the given runnable strictly after all Runnables that were submitted
    * before it, and using the {@code executor} passed to the constructor.     .
    */
   @Override
   public void execute(Runnable r) {
-    Preconditions.checkNotNull(r, "'r' must not be null.");
-    boolean scheduleTaskRunner = false;
-    synchronized (internalLock) {
-      waitQueue.add(r);
+    runQueue.offer(checkNotNull(r, "'r' must not be null."));
+    schedule();
+  }
 
-      if (!isThreadScheduled) {
-        isThreadScheduled = true;
-        scheduleTaskRunner = true;
-      }
-    }
-    if (scheduleTaskRunner) {
-      boolean threw = true;
+  private void schedule() {
+    if (running.compareAndSet(false, true)) {
+      boolean success = false;
       try {
-        executor.execute(taskRunner);
-        threw = false;
+        executor.execute(this);
+        success = true;
       } finally {
-        if (threw) {
-          synchronized (internalLock) {
-            // It is possible that at this point that there are still tasks in
-            // the queue, it would be nice to keep trying but the error may not
-            // be recoverable.  So we update our state and propogate so that if
-            // our caller deems it recoverable we won't be stuck.
-            isThreadScheduled = false;
-          }
+        // It is possible that at this point that there are still tasks in
+        // the queue, it would be nice to keep trying but the error may not
+        // be recoverable.  So we update our state and propogate so that if
+        // our caller deems it recoverable we won't be stuck.
+        if (!success) {
+          running.set(false);
         }
       }
     }
   }
 
-  /**
-   * Task that actually runs the Runnables.  It takes the Runnables off of the
-   * queue one by one and runs them.  After it is done with all Runnables and
-   * there are no more to run, puts the SerializingExecutor in the state where
-   * isThreadScheduled = false and returns.  This allows the current worker
-   * thread to return to the original pool.
-   */
-  private class TaskRunner implements Runnable {
```

Comment:
I was trying to drop easy allocations.  Making it the same object avoids allocating a TaskRunner.  CLQ costs an allocation per add so I was trying to pay as much of that cost back in this class.